lib/deploy/data.ts

```diff
@@ -0,0 +1,296 @@
+
+/**
+ * Generate KubernetesApplication from project and goal invocation.
+ * This function used [[defaultDeploymentData]] to produce initial
+ * [[KubernetesApplication]] data.  It then merges in any object
+ * provided by `goalEvent.data[[[sdmPackk8]]]`, with the latter taking
+ * precedence.  Then, if `registration.applicationData` is truthy, it
+ * calls that function, passing the merged value of the Kubernetes
+ * application, and uses its return value as
+ * `goalEvent.data[[[sdmPackk8]]]`.
+ *
+ * @param k8Deploy Kubernetes deployment goal
+ * @param registration Goal registration/configuration for `k8Deploy`
+ * @param goalInvocation The Kubernetes deployment goal currently triggered.
+ * @return The SdmGoalEvent augmented with [[KubernetesApplication]] in the data
+ */
+export function generateKubernetesGoalEventData(
+    k8Deploy: KubernetesDeploy,
+    registration: KubernetesDeployRegistration,
+    goalInvocation: GoalInvocation,
+): Promise<SdmGoalEvent> {
+    return k8Deploy.sdm.configuration.sdm.projectLoader.doWithProject({ ...goalInvocation, readOnly: true }, async p => {
+
+        const { context, goalEvent } = goalInvocation;
+        const defaultData: any = {};
+        defaultData[sdmPackK8s] = await defaultKubernetesApplication(p, goalEvent, k8Deploy, context);
+
+        const slug = goalEventSlug(goalEvent);
+        let eventData: any;
+        try {
+            eventData = JSON.parse(goalEvent.data || "{}");
+        } catch (e) {
+            logger.warn(`Failed to parse goal event data for ${slug} as JSON: ${e.message}`);
+            logger.warn(`Ignoring current value of goal event data: ${goalEvent.data}`);
+            eventData = {};
+        }
+        const mergedData = _.merge(defaultData, eventData);
+
+        if (registration.applicationData) {
+            const callbackData = await registration.applicationData(mergedData[sdmPackK8s], p, k8Deploy, goalEvent);
+            mergedData[sdmPackK8s] = callbackData;
+        }
+
+        goalEvent.data = JSON.stringify(mergedData);
+        return goalEvent;
+    });
+}
+
+/**
+ * Fetch [[KubernetesApplication]] from goal event.  If the goal event
+ * does not contain Kubernetes application information in its data
+ * property, `undefined` is returned.  If the value of the goal event
+ * data cannot be parsed as JSON, an error is thrown.
+ *
+ * @param goalEvent SDM goal event to retrieve the Kubernetes application data from
+ * @return Parsed [[KubernetesApplication]] object
+ */
+export function getKubernetesGoalEventData(goalEvent: SdmGoalEvent): KubernetesApplication | undefined {
+    if (!goalEvent || !goalEvent.data) {
+        return undefined;
+    }
+    let data: any;
+    try {
+        data = JSON.parse(goalEvent.data);
+    } catch (e) {
+        e.message = `Failed to parse goal event data (${goalEvent.data}): ${e.message}`;
+        logger.error(e.message);
+        throw e;
+    }
+    if (!data[sdmPackK8s]) {
```

Comment:
That if statement can be removed, can't it?  `data[sdmPackK8s]` returns `undefined` if it doesn't exist, no?

Reply:
I believe that is correct.